src/v/storage/index_state.cc

```diff
@@ -124,4 +128,90 @@ iobuf index_state::checksum_and_serialize() {
     return serde_compat::index_state_serde::encode(*this);
 }
 
+void index_state::serde_write(iobuf& out) const {
+    using serde::write;
+
+    iobuf tmp;
+    write(tmp, bitflags);
+    write(tmp, base_offset);
+    write(tmp, max_offset);
+    write(tmp, base_timestamp);
+    write(tmp, max_timestamp);
+    write(tmp, relative_offset_index.copy());
+    write(tmp, relative_time_index.copy());
+    write(tmp, position_index.copy());
+
+    crc::crc32c crc;
+    crc_extend_iobuf(crc, tmp);
+    const uint32_t tmp_crc = crc.value();
+
+    // data blob + crc
+    write(out, std::move(tmp));
+    write(out, tmp_crc);
+}
+
+void read_nested(
+  iobuf_parser& in, index_state& st, const size_t bytes_left_limit) {
+    /*
+     * peek at the 1-byte version prefix. serde support starts with the version
+     * immediately following the on-disk deprecated version.
+     */
+    if (
+      serde::peek_version(in)
+      > serde_compat::index_state_serde::ondisk_version) {
+        // consume the full serde header from the input
+        const auto hdr = serde::read_header<index_state>(in, bytes_left_limit);
+
+        using serde::read_nested;
+
+        // data blog + crc
+        iobuf tmp;
+        uint32_t tmp_crc = 0;
+        read_nested(in, tmp, hdr._bytes_left_limit);
+        read_nested(in, tmp_crc, hdr._bytes_left_limit);
```

Comment:
@felixguendling I still haven't formed an intuition for `bytes_left_limit`. it's const, so it isn't getting modified as we read via calls to read_nested. so is it a limit that the read_nested asserts isn't crossed?